### PR TITLE
feat(git-guards): add pre-commit-sign-guard PreToolUse hook

### PR DESCRIPTION
**Pull request by JacobPEvans-personal**
## Summary

New PreToolUse hook that intercepts `git commit` and refuses unsigned commits before they're created locally. Closes the structural gap that allowed local config drift to silently produce unsigned commits.

## Behavior

Denies any `git commit` invocation when:
- `commit.gpgsign` is not `true`, OR
- `user.signingkey` is empty, OR
- The configured signing key is not loadable (`gpg --list-secret-keys` for GPG, file readable for SSH), OR
- The command includes `-c commit.gpgsign=false`

Allows two specific patterns:
- `git commit --amend --no-edit -S` — the rebase `--exec` resign pattern used to backfill historical unsigned PRs (`-S` forces signing regardless of config)
- `GIT_GUARD_SIGN_OVERRIDE=1` env var — for tests, mirroring the `GIT_GUARD_BRANCH_OVERRIDE` pattern in `main-branch-guard.py`

## Why now

JacobPEvans repos enforce `required_signatures` org-wide. The recent audit found 37 unsigned commits across 12 open PRs — most from local-Mac sessions where signing config was incomplete. This hook makes that failure mode impossible: the unsigned commit never gets created.

## Architecture reference

JacobPEvans/ai-assistant-instructions#607 — canonical signing rule documenting the four-context architecture.

## Companion PR

JacobPEvans/claude-code-routines#6 — cloud-routine side of the signing fix (bootstrap-signing.sh + BOT_SETUP.md).

## Test plan

- [x] Smoke-tested locally: hook allows `git status`, denies `git -c commit.gpgsign=false commit`, allows `git commit --amend --no-edit -S`, respects `GIT_GUARD_SIGN_OVERRIDE=1`
- [x] `python3 -m py_compile` clean
- [ ] Add unit tests under `tests/git-guards/pre-commit-sign-guard/` in a follow-up PR (pattern parity with existing `tests/git-guards/main-branch-guard/`)

### CODE REVIEW

File: git-guards/scripts/pre-commit-sign-guard.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    try:
+        result = subprocess.run(
+            ["git", "config", "--get", key],
+            capture_output=True, text=True, timeout=2,
+        )
+        if result.returncode == 0:
+            return result.stdout.strip()
+    except (OSError, subprocess.SubprocessError):
+        pass
+    return ""
```
[the remark this answers is not in the document]
Comment (JacobPEvans-personal) [reply git-guards/scripts/pre-commit-sign-guard.py]: Fixed in 9786da2 — `_git_config_get` now takes a `bool_type=True` kwarg that delegates to `git config --type=bool --get`, normalising yes/on/1/true (and case variants) to the canonical string "true". Adopted your suggested signature shape (with a small renaming of `type` → `bool_type` to avoid shadowing the built-in).

File: git-guards/scripts/pre-commit-sign-guard.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            "test, set GIT_GUARD_SIGN_OVERRIDE=1 in the environment."
+        )
+
+    # Check the effective signing config
+    gpgsign = _git_config_get("commit.gpgsign").lower()
+    signingkey = _git_config_get("user.signingkey")
+    fmt = (_git_config_get("gpg.format") or "openpgp").lower()
+
+    missing = []
+    if gpgsign != "true":
```
[the remark this answers is not in the document]
Comment (JacobPEvans-personal) [reply git-guards/scripts/pre-commit-sign-guard.py]: Fixed in 9786da2 — the `gpgsign` lookup now uses `bool_type=True`, so `yes`/`on`/`1` no longer trip the "missing" branch. Confirmed by the new `-c commit.gpgsign=true (truthy override)` test case.

File: git-guards/scripts/pre-commit-sign-guard.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        if m:
+            rest = m.group(2)
+            continue
+        m = re.match(r"^-c\s+\S+\s*(.*)", rest)
+        if m:
+            rest = m.group(1)
+            continue
+        break
+    # Match `commit` as the actual subcommand; reject `commit-tree`, `commit-graph`
+    return bool(re.match(r"^commit(\s|$)", rest))
```
[the remark this answers is not in the document]
Comment (JacobPEvans-personal) [reply git-guards/scripts/pre-commit-sign-guard.py]: Fixed in 9786da2 — replaced the regex-based parser with `shlex.split` in the new `_parse_git_invocation`, plus shlex-token-based `_is_git_commit` and `_commit_explicitly_signs`. Test cases for `git\tcommit`, `git    commit` (multi-space), and `git -c "user.name=Foo Bar" commit` all pass.

File: git-guards/scripts/pre-commit-sign-guard.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+def _commit_only_amends_no_edit(command: str) -> bool:
+    """`git commit --amend --no-edit -S` is the rebase --exec resign pattern.
+    Allow it through — even though `commit.gpgsign` may not be set, the
+    explicit -S flag forces signing on this command.
+    """
+    if "--amend" not in command:
+        return False
+    return bool(re.search(r"(^|\s)-S(\s|$)", command) or
+                re.search(r"(^|\s)--gpg-sign(\s|=|$)", command))
```
[the remark this answers is not in the document]
Comment (JacobPEvans-personal) [reply git-guards/scripts/pre-commit-sign-guard.py]: Fixed in 9786da2 — moved away from substring matching entirely. The new `_commit_explicitly_signs` iterates shlex tokens checking for `-S`, `--gpg-sign`, or `--gpg-sign=...`. Also dropped the `--amend` requirement: it was never actually checked against `--no-edit` (copilot caught the same mismatch), and the hook's job is verifying signing intent — `-S` alone is sufficient.

File: git-guards/scripts/pre-commit-sign-guard.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        return False
+    if fmt == "ssh":
+        try:
+            return os.access(os.path.expanduser(key_value), os.R_OK)
+        except OSError:
+            return False
+    # Default to GPG (git's `gpg.format` defaults to `openpgp`)
+    try:
+        result = subprocess.run(
+            ["gpg", "--list-secret-keys", key_value],
```
[the remark this answers is not in the document]
Comment (JacobPEvans-personal) [reply git-guards/scripts/pre-commit-sign-guard.py]: Fixed in 9786da2 — `gpg --list-secret-keys -- <key_value>` now uses `--` to separate the positional arg from flags. Same hardening applied to the deny-message hint that reproduces the command.

File: git-guards/scripts/pre-commit-sign-guard.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        sys.exit(0)
+
+    # `git commit --amend --no-edit -S` is signing explicitly via flag;
+    # let it through even if the global config is incomplete.
+    if _commit_only_amends_no_edit(command):
+        sys.exit(0)
+
+    # `git -c commit.gpgsign=false commit` would slip past a config check
+    # but be caught here — anything that explicitly disables signing is denied.
+    if re.search(r"-c\s+commit\.gpgsign\s*=\s*false", command):
```
[the remark this answers is not in the document]
Comment (JacobPEvans-personal) [reply git-guards/scripts/pre-commit-sign-guard.py]: Fixed in 9786da2 — the regex now matches all bool-false variants per git's parser: `(false|no|off|0|''|"")` (case-insensitive), with a key-name boundary check. New tests cover `=false`, `=no`, `=off`, `=0`, and `=FALSE`.

File: git-guards/scripts/pre-commit-sign-guard.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (JacobPEvans-personal) [reply git-guards/scripts/pre-commit-sign-guard.py]: Fixed in 9786da2 — renamed to `_commit_explicitly_signs` and dropped the `--amend` check. The criterion is now just "is `-S`/`--gpg-sign` present?". This still allows the original `git rebase --exec 'git commit --amend --no-edit -S'` resign pattern (it has `-S`), and also correctly allows plain `git commit -S` (which forces signing regardless of config). Verified by the "plain -S without --amend" test case.

File: git-guards/scripts/pre-commit-sign-guard.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JacobPEvans-personal) [reply git-guards/scripts/pre-commit-sign-guard.py]: Fixed in 9786da2 — `_parse_git_invocation` now returns `(rest_tokens, cwd)` where `cwd` is the value of any `-C <path>` global flag. Every `_git_config_get` call accepts an optional `cwd` kwarg that's passed as `subprocess.run(..., cwd=cwd)`. Verified by the new test case that creates a tmpdir with explicitly-empty signing config and confirms `git -C $tmpdir commit` is denied (proves the hook queried the tmpdir, not the test runner's cwd).

File: git-guards/scripts/pre-commit-sign-guard.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        )
+
+    # Check the effective signing config
+    gpgsign = _git_config_get("commit.gpgsign").lower()
+    signingkey = _git_config_get("user.signingkey")
+    fmt = (_git_config_get("gpg.format") or "openpgp").lower()
+
+    missing = []
+    if gpgsign != "true":
+        missing.append("commit.gpgsign is not set to true")
```
[the remark this answers is not in the document]
Comment (JacobPEvans-personal) [reply git-guards/scripts/pre-commit-sign-guard.py]: Fixed in 9786da2 — same fix as gemini's high-priority comment. `_git_config_get("commit.gpgsign", bool_type=True)` invokes `git config --type=bool --get commit.gpgsign`, returning canonical "true"/"false" regardless of the underlying form (`yes`/`on`/`1` etc).

File: git-guards/scripts/pre-commit-sign-guard.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+(missing GPG/SSH key, `commit.gpgsign=false`, signing key not loadable) used
+to silently produce unsigned commits that landed on PR branches and blocked
+merges. This hook makes the failure loud and immediate — `git commit` is
+denied at hook time when the signing config is incomplete.
+
+Architecture: ai-assistant-instructions/agentsmd/rules/git-signing.md
+Local config schema: ~/.config/nix-home/local.nix (per nix-home README)
+
+Test bypass: set GIT_GUARD_SIGN_OVERRIDE=1 (mirrors the
+GIT_GUARD_BRANCH_OVERRIDE pattern in main-branch-guard).
```
[the remark this answers is not in the document]
Comment (JacobPEvans-personal) [reply git-guards/scripts/pre-commit-sign-guard.py]: Fixed in 9786da2 — replaced the broken in-repo path reference with the actual GitHub URL: <https://github.com/JacobPEvans/ai-assistant-instructions/blob/main/agentsmd/rules/git-signing.md>. Same fix applied to the script's module-level docstring.

File: git-guards/scripts/pre-commit-sign-guard.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+    # `git commit --amend --no-edit -S` is signing explicitly via flag;
+    # let it through even if the global config is incomplete.
+    if _commit_only_amends_no_edit(command):
+        sys.exit(0)
+
+    # `git -c commit.gpgsign=false commit` would slip past a config check
+    # but be caught here — anything that explicitly disables signing is denied.
+    if re.search(r"-c\s+commit\.gpgsign\s*=\s*false", command):
+        deny(
```
[the remark this answers is not in the document]
Comment (JacobPEvans-personal) [reply git-guards/scripts/pre-commit-sign-guard.py:199]: Fixed in 9786da2 — added `test_pre_commit_sign_guard.py` (parity with `test_permission_guard.py`). 22 cases covering: non-Bash tool, non-git/non-commit, `commit-tree`/`commit-graph` (must NOT match), whitespace-robust parsing (`git\tcommit`, `git    commit`), `GIT_GUARD_SIGN_OVERRIDE` bypass, all `-S`/`--gpg-sign` variants, all `-c commit.gpgsign=<falsey>` variants, truthy `-c` pass-through, quoted `-c` values with spaces, and `-C <path>` propagation against a synthetic empty-config repo. All pass locally.

File: git-guards/scripts/pre-commit-sign-guard.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JacobPEvans-personal) [reply git-guards/scripts/pre-commit-sign-guard.py]: Fixed in 9786da2 — replaced the regex with `shlex.split` in `_parse_git_invocation`. Falls open on `ValueError` (returning `(None, None)`) consistent with the git-permission-guard pattern. `git\tcommit`, `git    commit`, and `git -c "user.name=Foo Bar" commit` now all parse correctly — verified by new test cases.

**Comment by JacobPEvans** [comment]
Closing this PR — the local hook is unnecessary.

The native solution achieves the same outcome with zero scripts:

1. **`commit.gpgsign=true` set by nix-home** (Phase 1 of the broader plan) — every git invocation on the local Mac signs by default. No way to silently fall back to unsigned without an explicit `-c commit.gpgsign=false` flag.

2. **Server-side `required_signatures` ruleset** (already enforced on every default branch, will be universal in Phase 4) — any unsigned commit is rejected at push. The author finds out immediately and re-signs.

The 200-line Python hook was redundant defense — it didn't catch anything the server doesn't already catch, and it added a substantial maintenance surface (shlex parsing edge cases, `-C` path propagation, bool config handling, GPG vs SSH branching, a test suite for all of the above).

Worktree branch deleted. The signing architecture continues without any local pre-commit hook — see the `git-signing` rule in JacobPEvans/ai-assistant-instructions for the full picture.
